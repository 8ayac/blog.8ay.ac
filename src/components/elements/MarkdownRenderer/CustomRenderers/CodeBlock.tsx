import React from 'react';
import { Prism } from 'react-syntax-highlighter';
import { atomDark } from 'react-syntax-highlighter/dist/cjs/styles/prism';

export const CodeBlock: React.FC<{ language: string; value: string }> = ({
  language,
  value,
}) => (
  <Prism
    style={atomDark}
    language={language}
    customStyle={{
      padding: '1.5rem 2rem',
      borderRadius: '5px',
      fontFamily: 'monospace',
      lineHeight: '1.1',
    }}
  >
    {value}
  </Prism>
);
