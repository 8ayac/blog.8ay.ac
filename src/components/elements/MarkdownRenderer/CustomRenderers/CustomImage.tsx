import { config } from '@/site.config';
import Image from 'next/image';
import React from 'react';
import urljoin from 'url-join';

export const CustomImage: React.FC<{
  src: string;
  alt?: string;
  title?: string;
}> = ({ src, alt, title }) => {
  const srcUrl = /^https?:\/\/.*/.test(src)
    ? src
    : urljoin(config.site.rootUrl, src);
  const { searchParams } = new URL(srcUrl);
  const width = searchParams.get('w');
  const height = searchParams.get('h');

  return (
    <>
      {width && height ? (
        <Image
          src={src}
          alt={alt}
          title={title}
          width={`${width}px`}
          height={`${height}px`}
        />
      ) : (
        <img src={src} alt={alt} title={title} />
      )}
    </>
  );
};
