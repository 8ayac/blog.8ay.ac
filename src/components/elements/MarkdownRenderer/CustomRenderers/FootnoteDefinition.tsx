import {
  FOOTNOTE_DEFINITION_PREFIX,
  FOOTNOTE_REFERENCE_PREFIX,
} from '@/src/constants/footnote';
import styled from '@emotion/styled';
import React, { ReactElement } from 'react';

const S = {
  FootnoteWrapperDiv: styled.div`
    margin-bottom: 1rem !important;

    * {
      display: inline;
    }
  `,

  LabelSpan: styled.span`
    margin-right: 1rem;
    font-weight: 600;
  `,

  ReturnAnchor: styled.a`
    margin: 0 0.2rem !important;
  `,
};

export const FootnoteDefinition: React.FC<{
  identifier: string;
  label: string;
  children: ReactElement;
}> = ({ identifier, label, children }) => {
  return (
    <>
      <S.FootnoteWrapperDiv id={`${FOOTNOTE_DEFINITION_PREFIX}${identifier}`}>
        <S.LabelSpan>^{label}:</S.LabelSpan>
        {children}
        <S.ReturnAnchor href={`#${FOOTNOTE_REFERENCE_PREFIX}${identifier}`}>
          â†µ
        </S.ReturnAnchor>
      </S.FootnoteWrapperDiv>
    </>
  );
};
