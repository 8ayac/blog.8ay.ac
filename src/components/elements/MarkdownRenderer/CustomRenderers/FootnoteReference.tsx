import {
  FOOTNOTE_DEFINITION_PREFIX,
  FOOTNOTE_REFERENCE_PREFIX,
} from '@/src/constants/footnote';
import styled from '@emotion/styled';
import React from 'react';

const S = {
  IdentifierSup: styled.sup`
    margin: 0.2rem;
  `,
};

export const FootnoteReference: React.FC<{
  identifier: string;
  label: string;
}> = ({ identifier, label }) => {
  return (
    <S.IdentifierSup id={`${FOOTNOTE_REFERENCE_PREFIX}${identifier}`}>
      <a href={`#${FOOTNOTE_DEFINITION_PREFIX}${identifier}`}>{label}</a>
    </S.IdentifierSup>
  );
};
