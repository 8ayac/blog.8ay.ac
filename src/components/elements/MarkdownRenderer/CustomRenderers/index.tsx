import { CustomImage } from '@/src/components/elements/MarkdownRenderer/CustomRenderers/CustomImage';
import { FootnoteDefinition } from '@/src/components/elements/MarkdownRenderer/CustomRenderers/FootnoteDefinition';
import { FootnoteReference } from '@/src/components/elements/MarkdownRenderer/CustomRenderers/FootnoteReference';

export const customRenderers = {
  footnoteDefinition: FootnoteDefinition,
  footnoteReference: FootnoteReference,
  image: CustomImage,
};
