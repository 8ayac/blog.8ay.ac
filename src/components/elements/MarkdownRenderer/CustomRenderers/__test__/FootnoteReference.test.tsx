import { FootnoteReference } from '@/src/components/elements/MarkdownRenderer/CustomRenderers/FootnoteReference';
import {
  FOOTNOTE_DEFINITION_PREFIX,
  FOOTNOTE_REFERENCE_PREFIX,
} from '@/src/constants/footnote';
import { mount } from 'enzyme';
import React from 'react';

describe('FootnoteReference', () => {
  const testProps = {
    identifier: '1',
    label: '1',
  };
  const wrapper = mount(
    <FootnoteReference
      identifier={testProps.identifier}
      label={testProps.label}
    />
  );

  describe('is rendered correctly', () => {
    test('to match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('to be linked properly', () => {
      expect(wrapper.find('a').prop('href')).toBe(
        `#${FOOTNOTE_DEFINITION_PREFIX}1`
      );
    });

    test('to be specified proper id attribute', () => {
      expect(wrapper.find('sup').prop('id')).toBe(
        `${FOOTNOTE_REFERENCE_PREFIX}1`
      );
    });
  });

  describe('has proper style rules', () => {
    test('in IdentifierSup', () => {
      expect(wrapper.find('IdentifierSup').find('sup')).toHaveStyleRule(
        'margin',
        '0.2rem'
      );
    });
  });
});
