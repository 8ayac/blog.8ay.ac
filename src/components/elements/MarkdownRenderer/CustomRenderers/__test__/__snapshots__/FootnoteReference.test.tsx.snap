// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FootnoteReference is rendered correctly to match the snapshot 1`] = `
.emotion-0 {
  margin: 0.2rem;
}

<FootnoteReference
  identifier="1"
  label="1"
>
  <IdentifierSup
    id="fnref:1"
  >
    <sup
      className="emotion-0 emotion-1"
      id="fnref:1"
    >
      <a
        href="#fn:1"
      >
        1
      </a>
    </sup>
  </IdentifierSup>
</FootnoteReference>
`;
